### PR TITLE
[POC] wasm middlewares

### PR DESCRIPTION
**Pull request by gernest**
```json
{
    "name": "API with wasm middleware",
    "api_id": "3",
    "org_id": "default",
    "use_keyless": true,
    "proxy": {
        "listen_path": "/wasm/",
        "target_url": "http://httpbin.org/",
        "strip_listen_path": true
    },
    "do_not_track": true,
    "custom_middleware": {
        "pre": [
            {
                "name": "replace test header",
                "path": "http_headers.wasm"
            }
        ],
        "driver": "wasm"
    }
}
```

### CODE REVIEW

**Review by matiasinsaurralde — COMMENTED** [review]
Nice one!

File: wasm/v1/handler/http_request.go [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+	}
+	return &Header{head: r.Request.Trailer}
+}
+
+func (r *Request) GetHttpRequestMetadata() common.HeaderMap {
+	if r.Request == nil {
+		return nil
+	}
+	return &Header{head: r.Request.Trailer}
+}
```
Comment (asoorm) [inline wasm/v1/handler/http_request.go:45]: This looks like a bug.

**Comment by buger** [comment]
@gernest wasm package looks like a copy from another project. What was the original?

**Comment by gernest** [comment]
@buger https://github.com/gernest/tt

**Comment by gernest** [comment]
With exception of https://github.com/gernest/tt/blob/master/wasm/instance.go which Is a modified copy of https://github.com/mosn/proxy-wasm-go-host/blob/main/wasmer/instance.go

I will add proper attribution (Its under Apache). My code is under MIT.

**Comment by gernest** [comment]
My project has different goals and release cadence. I had to copy over so as to have something tyk specific for this POC.

**Comment by buger** [comment]
@gernest how extending it with custom function will look like? 
For example, exposing simple memory storage functions, SetKey, GetKey ?
Also how user can get access to user session object?

**Comment by gernest** [comment]
@buger 
>  how extending it with custom function will look like?
easy, you can register go functions that will be available in the module as imports.

> For example, exposing simple memory storage functions, SetKey, GetKey ?
This is already described  by the ABI so it is doable(I didn't implement this on this branch)

> Also how user can get access to user session object?
Many ways, they will require serialization/deserialrization though. we can extend and add functions for manipulating sessions if need be or we can stick with the ABI and expose using abi primitives.

**Comment by buger** [comment]
I do not think we need to close it

**Comment by gernest** [comment]
My bad! Is there still interest in this?

**Comment by excieve** [comment]
Sounds like a better way to do close to native speed plugins for most users. What is it missing to go from a PoC to a feature @gernest?

FYI @alephnull: that's the PoC I mentioned.

**Pull request closed without merging**